distributed/shuffle/shuffle_extension.py

```diff
@@ -450,44 +365,103 @@ class ShuffleSchedulerExtension:
 
     def __init__(self, scheduler):
         self.scheduler = scheduler
-        self.shuffles = defaultdict(lambda: defaultdict(dict))
+        self.scheduler.handlers.update(
+            {
+                "shuffle_get": self.get,
+            }
+        )
+        self.heartbeats = defaultdict(lambda: defaultdict(dict))
+        self.worker_for = dict()
+        self.schemas = dict()
+        self.columns = dict()
+        self.workers = dict()
 
     def heartbeat(self, ws, data):
         for shuffle_id, d in data.items():
-            self.shuffles[shuffle_id][ws.address].update(d)
+            self.heartbeats[shuffle_id][ws.address].update(d)
+
+    def get(
+        self, id: ShuffleId, schema: bytes | None, column, npartitions: int | None
+    ) -> dict:
+        if id not in self.worker_for:
+            assert schema is not None
+            assert column is not None
+            assert npartitions is not None
+            workers = list(self.scheduler.workers)
+            partitions = list(range(npartitions))  # TODO: check output partitions
+
+            # TODO: honor pre-existing restrictions
+            # TODO: only ask for desired outputs
+            mapping = {
+                part: worker_for(part, workers, npartitions) for part in partitions
+            }
+
+            name = "shuffle-unpack-" + id  # TODO single-source task name
+
+            for partition, worker in mapping.items():
+                key = f"('{name}', {partition})"
+                ts = self.scheduler.tasks[key]
+                if ts._worker_restrictions:
+                    raise NotImplementedError(
+                        "Shuffling and restrictions don't yet mix, use shuffle='tasks'"
+                    )
+                ts._worker_restrictions = {worker}
+
+            self.worker_for[id] = mapping
+            self.schemas[id] = schema
+            self.columns[id] = column
+            self.workers[id] = workers
+
+        return {
+            "worker_for": self.worker_for[id],
+            "workers": self.workers[id],
+            "column": self.columns[id],
+            "schema": self.schemas[id],
+        }
```

Comment:
So, we might regret pulling this into the Scheduler.  As the number of tasks climbs into the 100,000s this output may climb into the megabytes, which we then have to send to every active worker.  That *might* become an issue.

I'm inclined to wait until it does though before changing this.  There are two interventions:

1. Send a Pandas Series rather than a dict
2. Do this on a worker and use the worker networks to spread the data around